### PR TITLE
Do not attempt to pad nested tensors

### LINKED ISSUE #2040: pad_across_processes not compatible with PyTorch's nested_tensor
Buddy issue for https://github.com/huggingface/transformers/issues/26687 . The issue is basically identical in accelerate to transformers, so this issue can just be used for tracking and not discussion.

### PR DESCRIPTION
**Pull request by frankier**
# What does this PR do?

Fixes pad_across_processes erroring on nested tensors when reaching `len(tensor.shape)`

Fixes #2040

Looking for a bit of guidance/direction on whether other things are needed.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@pacman100

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@frankier Could you please merge with/rebase on main, that should fix the issue with CI.

**Comment by frankier** [comment]
> AFAICT, accelerate doesn't really take them into account for now, I wouldn't be surprised if more errors can occur. 

There are not more issues when I'm using accelerate from `transformers` Trainer, but there are issues in `transformers` if they are not converted in a custom `preprocess_logits_for_metrics`. Once this PR is merged, I will write another PR to transformers to improve the error messages in `transformers`.

> It would probably require some unit tests to check this.

The test I know I how to write now at the moment is just a smoke/unit test that checks that nested_tensor is returned as-is roundtripped when passed to `pad_across_processes`. I think this test would be a bit vacuous, but I can easily add it to the PR if you like. Could you please direct me which file/class would be the best place for it?

> As for padding, I don't understand yet why [to_padded_tensor](https://pytorch.org/docs/stable/nested.html#torch.nested.to_padded_tensor) cannot be used

The problem is that it uses in-band signalling. Whichever padding element you choose, may actually occur in the array. Adding this as default behavior could cause confusing bugs for users. It's better to force them to choose how to deal with the nested_tensor.

> what is the relation to numpy?

This might be from and old, possibly miscorrect memory I have about the default behavior of `preprocess_logits_for_metrics` converting everything to numpy arrays.

**Comment by BenjaminBossan** [comment]
> There are not more issues when I'm using accelerate from `transformers` Trainer, but there are issues in `transformers` if they are not converted in a custom `preprocess_logits_for_metrics`.

There could still be edge cases that you don't encounter but that other users may run into. E.g. the simple fact that `tensor.shape` may fail could easily lead to problems. Maybe accelerate dodges all of them, but we cannot know until we really test it. As is, I'd say we don't officially support nested tensors -- the docs mention they're at prototype stage, so it makes sense to wait.

That being said, there is no harm in accelerate getting out of the way if users happen to choose nested tensors, which is why I think this PR still makes sense.

> The test I know I how to write now at the moment is just a smoke/unit test that checks that nested_tensor is returned as-is roundtripped when passed to `pad_across_processes`. I think this test would be a bit vacuous, but I can easily add it to the PR if you like.

I don't think it would be vacuous. You need to consider that someone else might come in the future and remove the check inadvertently. Having a test would alert us to this.

> Could you please direct me which file/class would be the best place for it?

Not completely sure, [`test_utils.py`](https://github.com/huggingface/accelerate/blob/main/tests/test_utils.py) seems most appropriate, @muellerzr WDYT?

> The problem is that it uses in-band signalling. Whichever padding element you choose, may actually occur in the array.

I see, that makes sense. However, I do wonder if we should give a warning in that situation, so that the user is aware that padding did not take place.

> This might be from and old, possibly miscorrect memory I have about the default behavior of `preprocess_logits_for_metrics` converting everything to numpy arrays.

AFAICT, there is nothing in accelerate that requires this. A quick search didn't reveal anything in transformers `Trainer` either, but I don't have experience with it, so might be wrong.

**Comment by muellerzr** [comment]
Agree with all of @BenjaminBossan's points

> Not completely sure, [test_utils.py](https://github.com/huggingface/accelerate/blob/main/tests/test_utils.py) seems most appropriate, @muellerzr WDYT?

Yes indeed!

**Comment by frankier** [comment]
Please excuse the delay. I think I've addressed all this now.

**Comment by BenjaminBossan** [comment]
@frankier Could you please run `make style` and `make quality`?

**Comment by frankier** [comment]
Whoops. Thought I had(!) Done now.

**Review by BenjaminBossan — APPROVED** [review]
Thanks, looks good to me.

**Review by muellerzr — APPROVED** [review]
Thanks!
